executor/memtable_reader.go

```diff
@@ -997,3 +997,116 @@ func (e *hotRegionsHistoryRetriver) getHotRegionRowWithSchemaInfo(
 	}
 	return row, nil
 }
+
+type tikvRegionPeersRetriever struct {
+	dummyCloser
+	extractor *plannercore.TikvRegionPeersExtractor
+	retrieved bool
+}
+
+func (e *tikvRegionPeersRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest || e.retrieved {
+		return nil, nil
+	}
+	e.retrieved = true
+	tikvStore, ok := sctx.GetStore().(helper.Storage)
+	if !ok {
+		return nil, errors.New("Information about hot region can be gotten only when the storage is TiKV")
+	}
+	tikvHelper := &helper.Helper{
+		Store:       tikvStore,
+		RegionCache: tikvStore.GetRegionCache(),
+	}
+
+	if len(e.extractor.StoreIDs) < 1 && len(e.extractor.RegionIDs) < 1 {
+		regionsInfo, err := tikvHelper.GetRegionsInfo()
+		if err != nil {
+			return nil, err
+		}
+		return e.packTiKVRegionPeersRows(regionsInfo.Regions)
+	}
+
+	var regionsInfo, regionsInfoByStoreID, regionsInfoByRegionID []helper.RegionInfo
+	regionMap := make(map[int64]struct{})
+	if len(e.extractor.StoreIDs) > 0 {
+		for _, storeID := range e.extractor.StoreIDs {
+			regionsInfo, err := tikvHelper.GetStoreRegionsInfo(storeID)
+			if err != nil {
+				return nil, err
+			}
+			// remove dup region
+			for _, region := range regionsInfo.Regions {
+				if _, ok := regionMap[region.ID]; !ok {
+					regionMap[region.ID] = struct{}{}
+					regionsInfoByStoreID = append(regionsInfoByStoreID, region)
+				}
+			}
+		}
+		if len(e.extractor.RegionIDs) < 1 {
+			return e.packTiKVRegionPeersRows(regionsInfoByStoreID)
+		}
+	}
+
+	if len(e.extractor.RegionIDs) > 0 {
+		for _, regionID := range e.extractor.RegionIDs {
+			regionInfo, err := tikvHelper.GetRegionInfoByID(regionID)
+			if err != nil {
+				return nil, err
+			}
+			regionsInfoByRegionID = append(regionsInfoByRegionID, *regionInfo)
+		}
+		if len(e.extractor.StoreIDs) < 1 {
+			return e.packTiKVRegionPeersRows(regionsInfoByRegionID)
+		}
+	}
+
+	// intersect
+	for _, region := range regionsInfoByRegionID {
+		if _, ok := regionMap[region.ID]; ok {
+			regionsInfo = append(regionsInfo, region)
+		}
+	}
+	return e.packTiKVRegionPeersRows(regionsInfo)
```

Comment:
(3/3) We can intersect both `regionsInfoByStoreID` and `regionsInfoByRegionID` here, which would make the code more easy and clear.

Reply:
Ditto.